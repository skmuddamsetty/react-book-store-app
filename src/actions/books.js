// import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_BOOK
export const addBook = book => ({
  type: 'ADD_BOOK',
  book
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id
});
// EDIT_BOOK

export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates
});

export const startAddBook = (bookData = {}) => {
  return dispatch => {
    const { description = '', note = '', price = 0, createdAt = 0 } = bookData;
    const book = { description, note, price, createdAt };
    return database
      .ref('books')
      .push(book)
      .then(ref => {
        dispatch(addBook({ id: ref.key, ...book }));
      })
      .catch();
  };
};
