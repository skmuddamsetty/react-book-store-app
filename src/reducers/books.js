// Books reducer
const booksReducerDefaultState = [];

export const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      //   return state.concat(action.book);
      return [...state, action.book];
    case 'REMOVE_BOOK':
      return state.filter(book => book.id !== action.id);
    case 'EDIT_BOOK':
      return state.map(book => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    default:
      return state;
  }
};
