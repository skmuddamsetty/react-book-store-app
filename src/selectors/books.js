// get Visible Books
export const getVisibleBooks = (
  books,
  { text, sortBy, startDate, endDate }
) => {
  return books
    .filter(book => {
      const startDateMatch =
        typeof startDate !== 'number' || book.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== 'number' || books.createdAt <= endDate;
      const textMatch = book.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'price') {
        return a.price < b.price ? -1 : 1;
      } else {
        return 0;
      }
    });
};
