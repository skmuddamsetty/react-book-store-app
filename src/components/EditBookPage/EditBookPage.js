import React, { Component } from 'react';
import { connect } from 'react-redux';
import BookForm from '../BookForm/BookForm';
import { editBook, removeBook } from '../../actions/books';

export class EditBookPage extends Component {
  onEditBook = book => {
    this.props.editBook(this.props.book.id, book);
    this.props.history.push('/');
  };

  onRemoveBook = () => {
    this.props.removeBook({ id: this.props.book.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <BookForm book={this.props.book} onSubmit={this.onEditBook} />
        <button onClick={this.onRemoveBook}>Remove</button>
      </div>
    );
  }
}

// const EditBookPage = props => {
//   return (
//     <div>
//       <BookForm
//         book={props.book}
//         onSubmit={book => {
//           props.dispatch(editBook(props.book.id, book));
//           props.history.push('/');
//         }}
//       />
//       <button
//         onClick={() => {
//           props.dispatch(removeBook({ id: props.book.id }));
//           props.history.push('/');
//         }}
//       >
//         Remove
//       </button>
//     </div>
//   );
// };
const mapStateToProps = (state, props) => {
  console.log(state.books, props.match.params.id);
  return {
    book: state.books.find(book => book.id === props.match.params.id)
  };
};
const mapDispatchToProps = (dispatch, props) => {
  return {
    editBook: (id, book) => dispatch(editBook(id, book)),
    removeBook: id => dispatch(removeBook(id))
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditBookPage);
