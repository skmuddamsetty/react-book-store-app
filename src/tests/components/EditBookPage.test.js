import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { EditBookPage } from '../../components/EditBookPage/EditBookPage';
import { books } from '../fixtures/books';

let editBook, removeBook, history, wrapper;

beforeEach(() => {
  editBook = jest.fn();
  removeBook = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditBookPage
      editBook={editBook}
      removeBook={removeBook}
      history={history}
      book={books[1]}
    />
  );
});

test('should render EditBookPage properly', () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should handle editBook', () => {
  wrapper.find('BookForm').prop('onSubmit')(books[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editBook).toHaveBeenLastCalledWith(books[1].id, books[1]);
});

test('should handle removeBook', () => {
  wrapper.find('button').prop('onClick')(books[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeBook).toHaveBeenLastCalledWith({ id: books[1].id });
});

test('should handle removeBook using simulate click', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeBook).toHaveBeenLastCalledWith({ id: books[1].id });
});
