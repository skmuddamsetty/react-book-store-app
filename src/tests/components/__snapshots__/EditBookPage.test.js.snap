// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditBookPage properly 1`] = `
<div>
  <BookForm
    book={
      Object {
        "createdAt": 1,
        "description": "Angular Book",
        "id": "2",
        "note": "Angular Book Notes",
        "price": 10,
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
